prov/sm2/src/sm2_coordination.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _SM2_COORDINATION_H_
+#define _SM2_COORDINATION_H_
+
+#include "config.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/un.h>
+
+#include <ofi_atom.h>
+#include <ofi_hmem.h>
+#include <ofi_mem.h>
+#include <ofi_proto.h>
+#include <ofi_rbuf.h>
+#include <ofi_tree.h>
+
+#include <rdma/providers/fi_prov.h>
+
+#define SM2_NAME_MAX	      256
+#define SM2_INJECT_SIZE	      4096
+#define SM2_MAX_UNIVERSE_SIZE 4096
+/* TODO: Make the number of XFER ENTRY's configurable */
+#define SM2_NUM_XFER_ENTRY_PER_PEER 2048
```

Comment:
The maximum possible SHM file usage is 1MB + 16MB * 4096 = 65 GB, but this should not ever happen in typical situations.

We only expand the file as needed given the number of active endpoints on the system, and we do our best to clean up on exit and startup whenever we can.  If all processes exit cleanly, we'll never have more than the absolute minimum file size.

Reply:
switched SM2_MAX_UNIVERSE_SIZE to be 2048
switched SM2_NUM_XFER_ENTRY_PER_PEER to be 1024 (and one day we will make configurable).

we were getting a lot of bloat from:
```
size_t sm2_calculate_size_offsets(ptrdiff_t *rq_offset, ptrdiff_t *fs_offset)
{
	size_t total_size;

	total_size = sizeof(struct sm2_region);

	if (rq_offset)
		*rq_offset = total_size;
	total_size += sizeof(struct sm2_fifo);

	if (fs_offset)
		*fs_offset = total_size;
	total_size += freestack_size(sizeof(struct sm2_xfer_entry),
				     SM2_NUM_XFER_ENTRY_PER_PEER);

	/*
	 * TODO: Revisit later to see if we really need the size adjustment,
	 * or at most align to a multiple of a page size.
	 */
	total_size = roundup_power_of_two(total_size);

	return total_size;
}
```
I deleted that roundup_power_of_two and saw a performance enhancement, and our region size is now 4.24564MB

max shm size = 8.69476 gigabytes = 1MB + 4.245MB * 2048

Reply:
can we make it divisible by some number at least?  I'm tempted to say 4k (or whatever we assume for page size) but maybe at least 128 or maybe 512 byte alignment in case there is some vectorized alignment we need to use later.

Reply:
I would prefer if we play around with this after the main patch gets merged as follow on optimization.